ejb/WFLY-8449_Return_context_data_to_client.adoc

```diff
@@ -0,0 +1,81 @@
+= Return Context Data to EJB Clients
+:author:            Stuart Douglas
+:email:             sdouglas@redhat.com
+:toc:               left
+:icons:             font
+:keywords:          comma,separated,tags
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+This will allow for some data that has been placed in the InvocationContext by an EJB or Interceptor to
+be returned to the client so it can be processed by client interceptors. Only data that has explicitly
+been requested by the interceptor will be returned.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-8449[WFLY-8449]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/EAP7-976[EAP7-976]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Affected Projects or Components
+
+ * Wildfly
+ * EJB Client
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+An `EJBClientInterceptor` will be able to request data that it is interested in by calling
+`org.jboss.ejb.client.EJBClientInvocationContext#addReturnedContextDataKey(String key)`. This tells the server that the
+client interceptor would like any data that is stored under the provided key in the context data map to be returned
+to the client.
```

Comment:
We might want to consider something more generic; if it is reasonable for one server-side interceptor to use context data to pass information between interceptors (but not the client), then surely it's reasonable for the client side to do the same (without passing to the server)?  Wouldn't we want the set of keys to be marked as generically transportable, instead of sending all keys from client to server but only some from server to client?

Reply:
One option that Brad suggested would be for the client to provide a regex that can match key names

Reply:
Using a regex makes sense.  I'm just saying that we should treat client-to-server the same way we're treating server-to-client.

Reply:
Having just one regex key would probably be substantially more efficient than a set in most cases, at any rate.

Reply:
+1 for regex.

Reply:
So the issue with regex is that all the known use cases so far basically involve and server and client side interceptor communicating by putting data in a specific key. I really don't want to add regex support because I really don't see a valid use case for a client interceptor to start trying to examine an undefined number of keys.

If a client really does have a use case for this then it is very easy to implement by simply installing a server side interceptor that does the inspection, and sends the relevant data to the client under a known key.

Regex is an idea that sounds good on the surface, but IMHO anything that is sent over the wire should be carefully considered, rather than just vacuumed up as a regex glob.

Reply:
If we just send one key, we'll eventually get a request to send multiple.  If we send a set, we'll eventually get a complaint that the set is too big.  I think a regex is our best shot so far at not having to revisit this issue in the future, and it has the advantage of being a simple string, which is pretty hard to mess up.

Reply:
The customer controls both the server and client interceptors that send and receive the data. The server interceptor can send what is needed, the client interceptor can receive it. If they really actually want to just sent everything they can just have the interceptor put a copy the context data map into the context data.

Reply:
The current behavior with client interceptors is: anything they put into the contextData is sent to the server. I have not really seen any users using more than 1 client interceptor.  The regex, I was mainly thinking if a user made heavy use of this, they may just want everything returned so a *.  I have not really seen any users sending contextData through multiple interceptors, the use cases are typically client <-> server.  And if I were writing an app, I would call contextData.remove() if I did not want it passed along.